gnuradio-runtime/python/gnuradio/gr/bindings/block_gateway.h

```diff
@@ -0,0 +1,115 @@
+ *
+ * The gateway provides access to all the gr::block routines.
+ */
+using gw_block_t = enum {
+    GW_BLOCK_GENERAL = 0,
+    GW_BLOCK_SYNC,
+    GW_BLOCK_DECIM,
+    GW_BLOCK_INTERP
+};
+
+class block_gateway : public gr::block
+{
+public:
+    block_gateway(const py::handle& py_handle,
+                  const std::string& name,
+                  gr::io_signature::sptr in_sig,
+                  gr::io_signature::sptr out_sig);
+
+    // gr::block_gateway::sptr
+    using sptr = std::shared_ptr<block_gateway>;
+
+    /*!
+     * Make a new gateway block.
+     * \param py_handle the pybind11 object with callback
+     * \param name the name of the block (Ex: "Shirley")
+     * \param in_sig the input signature for this block
+     * \param out_sig the output signature for this block
+     * \return a new gateway block
+     */
+    static sptr make(const py::object& py_handle,
+                     const std::string& name,
+                     gr::io_signature::sptr in_sig,
+                     gr::io_signature::sptr out_sig);
+
+    // Protected members of gr::block trampolined here for python blocks to use
+    void _add_item_tag(unsigned int which_output, const gr::tag_t& tag);
+
+    void _add_item_tag(unsigned int which_output,
+                       uint64_t abs_offset,
+                       const pmt::pmt_t& key,
+                       const pmt::pmt_t& value,
+                       const pmt::pmt_t& srcid);
+
+    std::vector<gr::tag_t>
+    _get_tags_in_range(unsigned int which_input, uint64_t abs_start, uint64_t abs_end);
+
+    std::vector<gr::tag_t> _get_tags_in_range(unsigned int which_input,
+                                              uint64_t abs_start,
+                                              uint64_t abs_end,
+                                              const pmt::pmt_t& key);
+
+    std::vector<gr::tag_t>
+    _get_tags_in_window(unsigned int which_input, uint64_t rel_start, uint64_t rel_end);
+
+    std::vector<gr::tag_t> _get_tags_in_window(unsigned int which_input,
+                                               uint64_t rel_start,
+                                               uint64_t rel_end,
+                                               const pmt::pmt_t& key);
+
+    /*******************************************************************
+     * Overloads for various scheduler-called functions
+     ******************************************************************/
+    void forecast(int noutput_items, gr_vector_int& ninput_items_required) override;
+
+    int general_work(int noutput_items,
+                     gr_vector_int& ninput_items,
+                     gr_vector_const_void_star& input_items,
+                     gr_vector_void_star& output_items) override;
+
+    bool start() override;
+
+    bool stop() override;
+
+    void set_msg_handler_pybind(const pmt::pmt_t& which_port, std::string& handler_name);
+
+private:
+    py::handle _py_handle;
```

Comment:
According to [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md),
> All class data members shall begin with d_foo

I don't know how widely the standards are followed (my personal style is the same as yours), but I thought I'd mention it, for completeness.

Reply:
I don't have a preference and it's only this way because that's what it was in the former `_impl` class, so I'll probably change it.